### PR TITLE
On return type `impl Trait` for block with no expr point at last semi

### LINKED ISSUE #54771: Poor Trailing Semicolon Error in -> impl Trait Function
The diagnostics that check whether the concrete return type of an `-> impl Trait` function meets the `: Trait` bound should special-case the error when the concrete type is `()` to suggest removing the semicolon. Without this, the errors for a misplaced semicolon are misleading:

Without `impl Trait`:
```rust
trait Bar {}
impl Bar for u8 {}
fn foo() -> u8 {
    5;
}
```
```
error[E0308]: mismatched types
 --> src/main.rs:3:16
  |
3 |   fn foo() -> u8 {
  |  ________________^
4 | |     5;
  | |      - help: consider removing this semicolon
5 | | }
  | |_^ expected u8, found ()
  |
  = note: expected type `u8`
             found type `()`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `playground`.
```

With `impl Trait`:
```rust
trait Bar {}
impl Bar for u8 {}
fn foo() -> impl Bar {
    5;
}
```
```
error[E0277]: the trait bound `(): Bar` is not satisfied
 --> src/main.rs:3:13
  |
3 | fn foo() -> impl Bar {
  |             ^^^^^^^^ the trait `Bar` is not implemented for `()`
  |
  = note: the return type of a function must have a statically known size

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `playground`.
```

### PR DESCRIPTION
**Pull request by estebank**
Partial solution, doesn't actually validate that the last statement in the function body can satisfy the trait bound, but it's a good incremental improvement over the status quo.

```
error[E0277]: the trait bound `(): Bar` is not satisfied
  --> $DIR/impl-trait-return-trailing-semicolon.rs:3:13
   |
LL | fn foo() -> impl Bar {
   |             ^^^^^^^^ the trait `Bar` is not implemented for `()`
LL |     5;
   |      - consider removing this semicolon
   |
   = note: the return type of a function must have a statically known size
```

Partially addresses #54771.

### CODE REVIEW

**Comment by estebank** [comment]
cc @cramertj

**Comment by cramertj** [comment]
Nice! This looks like a great improvement.

File: src/librustc/traits/error_reporting.rs [hunk 1]
```diff
@@ -647,6 +648,27 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
...
+                            ..
+                        })) = node {
+                            let body = self.tcx.hir().body(*body_id);
+                            if let hir::ExprKind::Block(blk, _) = &body.value.node {
+                                if decl.output.span().overlaps(span) && blk.expr.is_none() &&
+                                    "()" == &trait_ref.self_ty().to_string()
+                                {
+                                    // When encountering a method with a trait bound not satisfied
+                                    // in the return type with a body that has no return, suggest
+                                    // removal of semicolon on last statement.
```
Comment (zackmdavis) [inline src/librustc/traits/error_reporting.rs]: The "we don't actually check if the last expression satisfies the trait" limitation seems significant enough to deserve a `FIXME (Issue #54771)` comment line

**Review by zackmdavis — APPROVED** [review]
requested reference comment; otherwise, r=me

**Comment by Centril** [comment]
Failed in rollup (https://github.com/rust-lang/rust/pull/58426#issuecomment-463224514) due to combination with https://github.com/rust-lang/rust/pull/58204. @bors r-

**Comment by Centril** [comment]
Ping from triage @estebank, any thoughts on moving forward?

**Comment by estebank** [comment]
@Centril I'll fix this soon

**Pull request merged**